dags/pull_traffic_signals.py

```diff
@@ -0,0 +1,458 @@
+            if not datetime_format_dates:
+                return None
+            # Case 2: only one date string can be converted into a datetime object
+            elif len(datetime_format_dates) == 1:
+                formatted_date = datetime_format_dates[0].strftime("%Y-%m-%d")
+                return formatted_date
+            # Case 3: more than one date string can be converted into datetime objects
+            else:
+                latest_date = datetime_format_dates[0]
+                for date in datetime_format_dates[1:]:
+                    if date > latest_date:
+                        latest_date = date
+                formatted_date = latest_date.strftime("%Y-%m-%d")
+                return formatted_date
+
+def pull_lpi(conn):
+    
+    local_table='vz_safety_programs_staging.signals_cart'
+    url = "https://secure.toronto.ca/opendata/cart/traffic_signals/v3?format=json"
+    return_json = requests.get(url).json()
+    
+    rows = []
+
+    # column names in the PG table
+    col_names = ['asset_type','px','main_street','midblock_route','side1_street','side2_street','latitude','longitude','details','activation_date']
+
+    # attribute names in JSON dict
+    att_names = ['px', 'main', 'mid_block', 'side1', 'side2', 'lat', 'long', 'lpiComment']
+
+    # each "info" is all the properties of one LPI, including its coords
+    for obj in return_json:
+        if obj['leading_pedestrian_intervals'] == 1:
+
+            # temporary list of properties of one LPI to be appended into the rows list
+            one_lpi = []
+
+            one_lpi.append('Leading Pedestrian Intervals') # append the asset_name as listed in EC2
+
+            # append the values in the same order as in the table
+            for attr in att_names:
+                one_lpi.append(obj[attr])
+
+            # append the latest LPI implementation date of any of the four directions
+            one_lpi.append(lastest_imp_date(obj))
+
+            rows.append(tuple(one_lpi))
+    
+    # delete existing LPI and insert into the local table
+    insert_data(conn, col_names, rows, 'Leading Pedestrian Intervals')
+
+# ------------------------------------------------------------------------------
+''' 
+Pull Traffic Signal Data (All of them) into signals_cart, in case a traffic signal was not pulled before
+'''
+def in_signals_cart(conn, px_number):
+    
+    ''' 
+    Returns a boolean that indicates whether a px is already in vz_safety_programs_staging.signals_cart
+    
+    Params
+    conn: connection to PG server
+    px_number: the 'px' value in each json object from the source URL
+    '''
+    with conn:
+        with conn.cursor() as cur:
+            cur.execute("SELECT EXISTS(SELECT 1 FROM vz_safety_programs_staging.signals_cart WHERE asset_type = 'Traffic Signals' AND px = %s)", (px_number,))
+            return cur.fetchone()[0]
+
+def pull_traffic_signal(conn):
```

Comment:
A last request is to save traffic signal information pulled from OpenData in a separate (new) table in order to have an exact copy of what exists on OpenData. This should be in addition to what's already implemented of inserting selected columns into `signals_cart`

Reply:
we also need to audit that new table of traffic signals